_posts/2021-09-20-verifying-lists-using-assertj.md

```diff
@@ -0,0 +1,462 @@
+                        return new Session.SessionBuilder()
+                                .id(session.getId())
+                                .start(session.getStart())
+                                .end(session.getEnd())
+                                .workouts(session.getWorkouts())
+                                .person(null)
+                                .build();
+                    }
+                    return session;
+                }
+        ).collect(Collectors.toList());
+
+        assertThat(sessions)
+                .filteredOn(session -> session.getStart().isAfter(LocalDateTime.now().minusHours(1)))
+                .extracting("person.name")
+                .contains("Bruce","Carol","Natalia");
+    }
+```
+
+### Flatmap and Method Call Extracting
+
+Our previous example where we wanted to find persons that had workout sessions done today was complex to implement using predicates, for such use cases flatmap comes to help.
+
+Flatmap is a method for extracting child of an object, which is a list, and proceeding to do further assertions on that list. We can use flatmap in two ways. If you want to extract some property and you have a getter for that property, you can just state the name of that property:
+```
+    @Test
+    void filterOnAllSessionsThatAreFromToday_flatMapExtracting(){
+        assertThat(personService.getAll()).flatExtracting("sessions")
+                .filteredOn(session -> ((Session)session).getStart().isAfter(LocalDateTime.now().minusHours(1)))
+                .extracting("person.name")
+                .contains("Tony", "Carol","Bruce","Natalia");
+    }
+```
+On the other hand, if you want to create a real extractor that will do some logic on top of the extracted list, you can create it using Java functions:
+
+```
+  public class PersonExtractors {
+    public PersonExtractors(){}
+
+    public static Function<Person, List<Session>> sessions(){
+        return new PersonSessionExtractor();
+    }
+
+    private static class PersonSessionExtractor implements Function<Person, List<Session>> {
+        @Override
+        public List<Session> apply(Person person) {
+            return person.getSessions();
+        }
+    }
+  }
+```
+And call it like this:
+```
+    @Test
+    void filterOnAllSessionsThatAreFromToday_flatMapExtractingMethod(){
+        assertThat(personService.getAll()).flatExtracting(PersonExtractors.sessions())
+                .filteredOn(session -> session.getStart().isAfter(LocalDateTime.now().minusHours(1)))
+                .extracting("person.name")
+                .contains("Tony", "Carol","Bruce","Natalia");
+    }
+```
+Method call extracting is similar to flatmap extracting. It allows us to call the method on top of each object in the list and create a new list on it, which we will continue our assertion. Simply call __.extractResultOf(“method name”)__ and continue with the assertion as needed.
+
+## Soft Assertions
```

Comment:
Soft assertions don't really have anything to do with list assertions, do they?

They would fit better into a separate article titled "AssertJ productivity tricks" or similar. I would remove it from this article for now, to focus only on lists.